DESIGN.md

```diff
@@ -0,0 +1,173 @@
+
+### Starting processes
+
+Example command:
+
+```console
+client $COMMON_OPTIONS start echo 'Hello, world!'
+```
+
+Any arguments after `start` are passed to the server. `--` is frequently also
+used by convention in cases like this to indicate "end of options", but in this
+case `start` will also fulfill this role and not requiring `--` is less typing.
+
+Once the server has started the process, the client will print out the ID that
+was assigned by the server to the new process and then exit.
+
+If the client has insufficient permissions to start a process, an error message
+will be printed indicating that this is the case.
+
+### Checking the status of a process
+
+Example command:
+
+```console
+client $COMMON_OPTIONS status 4
+```
+
+The `status` subcommand takes the process ID to query as its single argument.
+
+* If the process is still running, it will print `running`.
+* If the process has exited normally, it will print `exited: 0` (where `0` will
+  be the process's exit code).
+* If the process was stopped by a client's request, it will print `stopped`.
+  (See also `func (WaitStatus) Signaled`.)
+* If the process does not exist, it will print `process not found`.
+
+### Stopping a process
+
+Example command:
+
+```console
+client $COMMON_OPTIONS stop 4
+```
+
+The `stop` subcommand takes the process ID to stop as its single argument.
+
+* If the process ID does exist, it will print `process stopped` once the process
+  has been stopped.
+* If the process ID does not exist, it will print `process not found`.
+* If the client has insufficient permissions to stop a process, an error
+  message will be printed indicating that this is the case.
+
+Ideally, there would be a system for gracefully stopping processes by sending
+SIGINT, SIGTERM, and SIGKILL in that order over some time interval if the
+process did not stop quickly enough. This will not be implemented in the
+interest of time, processes will be sent a single signal and will be expected to
+terminate as a result.
+
+### Streaming the output of a process
+
+Example command:
+
+```console
+client $COMMON_OPTIONS stream 4
+```
+
+The `stream` subcommand takes the process ID whose output to stream as its
+single argument. `stdout` and `stderr` of the remote process will be forwarded
+to `stdout` and `stderr` of the client process respectively. The output of the
+remote process will be sent to the client since the process was started rather
+than since the client began streaming the output.
+
+If the process does not exist, it will print `process not found`.
+
+## Implementation details
+
+### mTLS
+
+The client and server will use mTLS to authenticate each other, and to allow
+the server to authorize the client. Clients with unknown certificates will be
+rejected by the server.
+
+The minimum supported TLS version shall be 1.3, which in turn requires a strong
+set of cipher suites.
+
+### Authorization
+
+Clients will have two authorization levels:
```

Comment:
Above it's mentioned multiple client certs can be configured, could you elaborate on how that works? Will they be used as fallbacks, or will the client need to have specific knowledge of which RPCs require which cert roles?